src/librustc_mir/build/mod.rs

```diff
@@ -383,6 +383,29 @@ fn construct_fn<'a, 'gcx, 'tcx, A>(hir: Cx<'a, 'gcx, 'tcx>,
     let source_info = builder.source_info(span);
     let call_site_s = (call_site_scope, source_info);
     unpack!(block = builder.in_scope(call_site_s, LintLevel::Inherited, block, |builder| {
+
+        // According to rustc_trans/callee we mark all foreign instances with nounwind,
+        // so let's make sure we don't unwind.
+        // Therefore generate an extra "Abort" landing pad.
+
+        // FIXME: Figure out why we can't use something like this instead:
+        // tcx.is_foreign_item(tcx.hir.local_def_id(fn_id));
+        // tcx.has_attr(tcx.hir.local_def_id(fn_id), "unwind");
+
+        let is_foreign = match abi {
```

Comment:
Actually, I think that an `abi` check is exactly the right thing. The key point is that the "C" ABI (and other non-Rust ABIs) don't have a defined way to propagate Rust panics, right?

Reply:
I'm not so sure.

"__rust_start_panic" and "panicking::rust_begin_panic" are of C ABI and are able to panic, so it can't be **that** undefined...

Rather, the real danger is when we tell LLVM that a function is "nounwind" and then we end up panicking within - or through - it. That's the undefined behavior this patch is trying to resolve. 
So, then I was trying to figure out when we actually mark a function as "nounwind", and it seems now I did not look closely enough. The algorithm seems to be:

 1) [ABI check](https://github.com/rust-lang/rust/blob/master/src/librustc_trans/declare.rs#L141) - so you're right, it should be an ABI check.
 2) Set as unwinding [if there is an unwind attribute](https://github.com/rust-lang/rust/blob/master/src/librustc_trans/attributes.rs#L120)
 3) Set as unwinding [if it isn't a foreign item](https://github.com/rust-lang/rust/blob/master/src/librustc_trans/callee.rs#L107-L115)

So maybe that's what I'm supposed to mimic, or possibly try to refactor somehow if we need it in two places?

Reply:
It seems like it would be ideal to have the criteria extracted into a helper function, yes.

Reply:
I'm not sure extracting the criteria *directly* would help:

1. Unwinding across a lang boundary is "instant" LLVM UB as we emit the `nounwind` attribute. If we want to continue doing that, we can't also make it trap.
2. Therefore, we want to catch unwinding *before* we reach the lang boundary. That means stopping unwinding from  proceeding from Rust to C, because we don't control the C-to-Rust lang boundary.
3. This means that we need to prevent unwinding on non-foreign items, which means we need to ignore the code for (3) from the previous list.

Reply:
I must be confused about something. @arielb1 what do you mean by this:

> Unwinding across a lang boundary is "instant" LLVM UB as we emit the nounwind attribute. If we want to continue doing that, we can't also make it trap.

Do you mean that the *call* is tagged with nounwind, or the function? I was assuming the latter.

Reply:
It doesn't matter whether the call or the function are tagged as nounwind. In both cases, unwinding is UB LLVM-side and therefore can't be turned to an abort..

Reply:
Sure enough. So for that we'll have to remove the "Set as unwinding if it isn't a foreign item" check.